src/machines/SyncMachine.ts

```diff
@@ -0,0 +1,304 @@
+              target: 'idle',
+            },
+            onError: `#${SYNC_MACHINE_STATES.ERROR}`,
+          },
+        },
+        [CONNECTED_STATES.handlingMetadataChanged]: {
+          id: 'handlingMetadataChanged',
+          initial: 'detectingDiff',
+          states: {
+            detectingDiff: {
+              invoke: {
+                src: 'metadataDiff',
+                onDone: { actions: ['metadataDecided'] },
+              },
+              on: {
+                METADATA_DECIDED: [
+                  { target: `#${CONNECTED_STATES.handlingVoidedTx}`, cond: 'metadataVoided', actions: ['unwrapEvent'] },
+                  { target: `#${CONNECTED_STATES.handlingUnvoidedTx}`, cond: 'metadataUnvoided', actions: ['unwrapEvent'] },
+                  { target: `#${CONNECTED_STATES.handlingVertexAccepted}`, cond: 'metadataNewTx', actions: ['unwrapEvent'] },
+                  { target: `#${CONNECTED_STATES.handlingFirstBlock}`, cond: 'metadataFirstBlock', actions: ['unwrapEvent'] },
+                  { target: `#${CONNECTED_STATES.handlingUnhandledEvent}`, cond: 'metadataIgnore' },
+                ],
+              },
+            },
+          },
+        },
+        // We have the unchanged guard, so it's guaranteed that this is a new tx
+        [CONNECTED_STATES.handlingVertexAccepted]: {
+          id: CONNECTED_STATES.handlingVertexAccepted,
+          invoke: {
+            src: 'handleVertexAccepted',
+            data: (_context: Context, event: Event) => event,
+            onDone: {
+              target: 'idle',
+              actions: ['sendAck', 'storeEvent', 'updateCache'],
+            },
+            onError: `#${SYNC_MACHINE_STATES.ERROR}`,
+          },
+        },
+        [CONNECTED_STATES.handlingVoidedTx]: {
+          id: CONNECTED_STATES.handlingVoidedTx,
+          invoke: {
+            src: 'handleVoidedTx',
+            data: (_context: Context, event: Event) => event,
+            onDone: {
+              target: 'idle',
+              actions: ['storeEvent', 'sendAck', 'updateCache'],
+            },
+            onError: `#${SYNC_MACHINE_STATES.ERROR}`,
+          },
+        },
+        [CONNECTED_STATES.handlingUnvoidedTx]: {
+          id: CONNECTED_STATES.handlingUnvoidedTx,
+          invoke: {
+            src: 'handleUnvoidedTx',
+            data: (_context: Context, event: Event) => event,
+            onDone: {
+              target: `#${CONNECTED_STATES.handlingVertexAccepted}`,
+              actions: ['storeEvent', 'sendAck', 'updateCache'],
+            },
+            onError: `#${SYNC_MACHINE_STATES.ERROR}`,
+          },
+        },
+        [CONNECTED_STATES.handlingFirstBlock]: {
+          id: CONNECTED_STATES.handlingFirstBlock,
+          invoke: {
+            src: 'handleTxFirstBlock',
+            data: (_context: Context, event: Event) => event,
+            onDone: {
+              target: 'idle',
+              actions: ['storeEvent', 'sendAck', 'updateCache'],
```

Comment:
Even though I agree with your point that it could be easy to change one of them and forget the others, I think I prefer being explicit about what actions are taken on each state migration so it's clear when reading the machine

Reply:
Ok, thanks!